node_modules/oae-authz/lib/api.js

```diff
@@ -545,14 +564,141 @@ var _explodeGroupMemberships = function(principalId, callback) {
             };
         });
 
-        // Update both the straight memberships cache and the dedicated indirect cache with the exploded memberships
+        // Update both the full memberships cache and the dedicated indirect cache with the exploded memberships
         Cassandra.runBatchQuery(_.union(authzMembershipsCacheQueries, authzMembershipsIndirectCacheQueries), function(err) {
             if (err) {
                 return callback(err);
             }
 
-            return callback(null, allMemberships, indirectMemberships);
+            return callback(null, graph);
+        });
+    });
+};
+
+/**
+ * Get the full members AuthzGroup for a collection of resources. The notion of "full" indicates
+ * that it will first get all direct users and groups that have access to the resource then
+ * recursively continue to gather all users and groups that have indirect access VIA other groups.
+ * The graph has the following guarantees:
+ *
+ *  * The nodes in the graph represent principals, except for any nodes representing the provided
+ *    `resourceIds` that are not groups (e.g., a members graph for a content item will have a node
+ *    representing the content item resource itself)
+ *  * The resulting graph will always contain the nodes representing the given `resourceIds`, even
+ *    if they have no members
+ *  * Each node has an `id` property which indicates the resource id (e.g., "c:cam:Foo.docx",
+ *    "u:oae:branden")
+ *  * The inbound edges of each node denote a "has member" relationship
+ *      * i.e., (Branden) --> (Foo.docx) indicates that the "Foo.docx" document "has member"
+ *        "Branden"
+ *  * The outbound edges then obviously denote a "is member of" relationship
+ *      * i.e., (Branden) --> (Foo.docx) indicates that "Branden" "is a member of" the "Foo.docx"
+ *        document
+ *  * Each edge has a `role` property which indicates the role the source principal has in the
+ *    target resource/principal
+ *
+ * @param  {String[]}       resourceIds     The ids of the resources whose full members graph to get
+ * @param  {Function}       callback        Standard callback function
+ * @param  {Object}         callback.err    An error that occurred, if any
+ * @param  {AuthzGraph}     callback.graph  The members graph of the given resource
+ */
+var getAuthzMembersGraph = module.exports.getAuthzMembersGraph = function(resourceIds, callback, _graph, _resourceIds) {
+    _graph = _graph || new AuthzGraph();
+    _resourceIds = _resourceIds || resourceIds.slice();
+    if (_.isEmpty(_resourceIds)) {
+        return callback(null, _graph);
+    }
+
+    // Get all the members of the given resource
+    var resourceId = _resourceIds.shift();
+    getAuthzMembers(resourceId, null, 10000, function(err, members) {
+        if (err) {
+            return callback(err);
+        }
+
+        // Add each member to the current members graph and associate the appropriate edges
+        _graph.addNode(resourceId);
```

Comment:
can you move `_graph.addNode(resourceId);` up to where you shift it out of the `resourceIds` array?